actionpack/lib/action_dispatch/journey/gtg/simulator.rb

```diff
@@ -22,18 +24,19 @@ def initialize(transition_table)
 
         def memos(string)
           input = StringScanner.new(string)
-          state = [0]
-          while sym = input.scan(%r([/.?]|[^/.?]+))
+          state = INITIAL_STATE
+
+          while sym = input.scan(MEMO_PATTERN)
             state = tt.move(state, sym)
           end
 
-          acceptance_states = state.find_all { |s|
-            tt.accepting? s
-          }
-
-          return yield if acceptance_states.empty?
+          acceptance_states = state.each_with_object([]) do |s, memos|
+            next unless tt.accepting? s
+            memo = tt.memo(s)
+            memos.concat(memo) unless memo.empty?
+          end
 
-          acceptance_states.flat_map { |x| tt.memo(x) }.compact
+          acceptance_states.empty? ? yield : acceptance_states
```

Comment:
Just to throw out a combined version of some different suggestions while we're here:

```ruby
def memos(string)
  state = INITIAL_STATE
  string.scan(%r([/.?]|[^/.?]+)) { |sym| state = tt.move(state, sym) } # Little dubious about this or its speed, but maybe.

  state.flat_map { |s| tt.memo(s) if tt.accepting? s }.presence or yield
end
```

Reply:
Just tried and using scan with a block is a little slower.
```
********************************************************************************
NOMATCH
********************************************************************************
IPS
Warming up --------------------------------------
               memos    41.530k i/100ms
          fast_memos    29.660k i/100ms
Calculating -------------------------------------
               memos    457.402k (± 4.6%) i/s -      2.284M in   5.003534s
          fast_memos    311.900k (± 5.7%) i/s -      1.572M in   5.054984s

Comparison:
               memos:   457402.2 i/s
          fast_memos:   311900.3 i/s - 1.47x  slower

MEMORY
Calculating -------------------------------------
               memos   496.000  memsize (     0.000  retained)
                         9.000  objects (     0.000  retained)
                         2.000  strings (     0.000  retained)
          fast_memos   656.000  memsize (     0.000  retained)
                        10.000  objects (     0.000  retained)
                         2.000  strings (     0.000  retained)

Comparison:
               memos:        496 allocated
          fast_memos:        656 allocated - 1.32x more
********************************************************************************
MATCH
********************************************************************************
IPS
Warming up --------------------------------------
               memos    37.205k i/100ms
          fast_memos    26.816k i/100ms
Calculating -------------------------------------
               memos    412.826k (± 4.3%) i/s -      2.083M in   5.056094s
          fast_memos    287.797k (± 6.5%) i/s -      1.448M in   5.051602s

Comparison:
               memos:   412826.0 i/s
          fast_memos:   287796.6 i/s - 1.43x  slower

MEMORY
Calculating -------------------------------------
               memos   456.000  memsize (     0.000  retained)
                         8.000  objects (     0.000  retained)
                         2.000  strings (     0.000  retained)
          fast_memos   616.000  memsize (     0.000  retained)
                         9.000  objects (     0.000  retained)
                         2.000  strings (     0.000  retained)

Comparison:
               memos:        456 allocated
          fast_memos:        616 allocated - 1.35x more
```